### DISCUSSION TITLE
Allow user to toggle between themes [light/dark]

### DISCUSSION THREAD

**Discussion by rajesh-s** [mmistakes/minimal-mistakes#2033] [Ideas]
## Summary
I'm new to Front-end development. I was wondering if it would be possible to allow the user to toggle between themes/css properties using the same color configurations specified using minimal_mistakes_skin: "default"

## Motivation

* In the recent times a lot of people prefer a darker background at night that's easier on the eyes
* At the same time, it's a know fact that using light font on a dark background ALWAYS reduces readability of the site. 
* I came across site like https://fdp.io/# that allow switching between themes using css properties
* I was wondering if there is a way to allow the user to switch between the colour themes

**Comment by ghost** [comment]
Like this? https://codepen.io/sashatran/pen/rPaLgG

**Comment by rajesh-s** [comment]
Yes that's right. Another example of it is the following Hugo theme https://academic-demo.netlify.com

**Comment by astrob0t** [comment]
This is how i _(a noob)_ achieved/reverse-engineered it in Minimal-Mistakes
* the Sass mixins were a tad bit confusing to me. So I extracted the css files by running the theme locally in default and dark mode. i saved them as `main.css` and `dark.css`
* pulled up kdiff3 and compared the key differences
* set up [CSS variables](https://www.w3schools.com/css/css3_variables.asp) for the differences
* started substituting the CSS colors in the `main.css` with the variables.
* added a `light` class to the `<body>` tag
* toggled that class on click of a button using javascript.

you can see it in action here _(shameless plug)_ - [my dev site](https://astrob0t.com)
There are a few other decorations on the site which are not directly from the dark theme but the above is the gist.

I am sure Michael will come up with a better solution than that of the above monstrosity. Eagerly waiting as this issue has been marked with the `Enhancement` label.

**Comment by mmistakes** [comment]
Not exactly sure how I'd handle this without bloating the theme with a bunch of extra CSS.
The way the stylesheet works now is at build it injects "skin" dependent Sass variables to give everything the correct colors.

Which means the dark skin's styling isn't available to the "light" or vice/versa. Does bundling in all the skins in a stylesheet really the best thing for those who don't care about a light/dark toggle? I dunno...

There's also the issue of the theme not having a canonical "light" skin anymore. There are essentially 10 different color palettes someone can choose... and they don't all compliment each other. Which means probably designing 9 or so more color schemes to invert them.

Personally I've never much cared for sites that add this sort of thing. I'd rather see it happen at the system or browser level. There are browser extensions that do just this which might have better results vs. me trying to maintain consistency across 10 color palettes x 2 modes (20 palettes if you're counting at home).

**Comment by ghost** [comment]
@mmistakes see https://inclusive-components.design/a-theme-switcher/

**Comment by ghost** [comment]
That said I see this as a passing fad. Everyone scrambling to add a light/dark mode to their themes when the OS + browser should be responsible for this behavior.

**Comment by astrob0t** [comment]
OS + browser can go all dark mode on their browser menu but they can only go as far as just applying a dark contrast mode on the website content, which may lead to the website visuals getting all messed up. So it is actually upto the website developer to manage/feature their own dark mode.

**Comment by ghost** [comment]
I agree it's up to the site to control how it looks, but feel the toggle needs to be elsewhere which it seems like macOS is moving wards with Mojave as illustrated by @paulmillr here: https://paulmillr.com/posts/using-dark-mode-in-css/

**Comment by rcarmo** [comment]
I was wondering if we could resume the discussion around Mojave dark mode, since I'm considering using this theme for docs and not having a dark/light toggle is a bit of a pain point (not insurmountable, just a very nice to have).

**Comment by jaketae** [comment]
I know this issue has been marked as closed, but I was wondering if we could continue this discussion as per @rcarmo's suggestion. I'm aware that there is no canonical light or dark theme, but it would be great if some form of customization functionality could be added with a toggling option, i.e. allow users to enable a toggle with two themes specified in `_config.yml`?

**Comment by Gioele-Bencivenga** [comment]
I'd also very much like to be able to achieve this.
Since my website is supposed to have both showcase pages and posts I'd like for users to be able to switch to the light theme when reading.

**Comment by hardianlawi** [comment]
@astrob0t Hi, I'm a noobie in web dev here. Do you mind sharing your solution in more details?

**Comment by khanfarhan10** [comment]
Sorry for not viewing this thread earlier,
@mmistakes my approach was a bit different : since you already have 10 different themes, would it be too difficult to add the toggle switch that switches between any 2 (or all :)) of them? A simple change in the css files would maybe do the trick as separately generated with the two different themes as suggested by @astrob0t ? Like if theme is dark use a dark.css and if theme is light use a light.css?

I'm a noob to jekyll and web dev so apologies if this is not easy at all. Great theme Michael. Will love it always!

Here is my original **ISSUE** :



### Summary
Wouldn't it be great to add a two theme switching button feature to the website:   
A button that would change the theme right away and would be available on all pages on the top navigation menu, possibly close to the search button on top?

Maybe include something in the theme configuration file as :
- Light_Dark_Button : true/ false
- Light_Theme : aqua
- Dark_Theme : neon



### Motivation
I thought of looking at some articles and found these : 
- https://medium.com/@derekkedziora/how-to-add-a-dark-mode-toggle-to-a-jekyll-site-a76dac00128d
- https://laptrinhx.com/how-to-add-a-dark-mode-toggle-to-a-jekyll-site-2605890340/

However, these documents don't cover the themes that minimal mistakes uses, aka : 

| aqua | neon | plum | dirt | mint | sunrise | air | contrast | dark |



### Drawbacks
None that I know of , except that it may cause the one navigation menu item to maybe go to the sidebar with three horizontal bars?

**Comment by sohamsaha99** [comment]
It'd be great to have this toggle feature.
I have also checked the two themes : "Basically Basic" and "So Simple". They are very nice. But in the current scenario, when many people spends a lot of time in the internet, (and a lot of them are night owls), allowing the user to choose between light and dark mode is a necessity for content developers.
I understand that there are total 10 themes. So an option for toggling between default and dark theme might be a great starting point.

**Comment by sohamsaha99** [comment]
I've been able to achieve this. Before telling the procedure, let's discuss the main idea:
There are going to be two skins. So first we define them in _config.yml file. Then we generate two css files for two skins. In the <head> tag, we include both of them, but only one is activated. Once the user clicks on the toggle button, the other css file is activated. Also we store the current theme in a variable, so that the skin remains consistent while we navigate to other pages.
Here is the procedure:

First, edit _config.yml : There are going to be two themes. The first one is declared as usual. And for the second one, we create a new entry caled minimal_mistakes_skin2. So, add the following lines:
```
minimal_mistakes_skin: "default"
minimal_mistakes_skin2: "dark"
```

Create a file in your project directory in the location assets/css/theme2.scss (create a folder named assets, and inside that, create another folder named css) and insert the following lines in the file:
```
---
# Only the main Sass file needs front matter (the dashes are enough)
---

@charset "utf-8";

@import "minimal-mistakes/skins/{{ site.minimal_mistakes_skin2 | default: 'default' }}"; // skin
@import "minimal-mistakes"; // main partials
```

Now, create a _includes folder and copy the _includes/head.html from your Minimal-mistakes gems directory. Now we have to modify the file. Open it and find the following line:
```
<link rel="stylesheet" href="{{ '/assets/css/main.css' | relative_url }}">
```
Change it to:
```
<link rel="stylesheet" href="{{ '/assets/css/main.css' | relative_url }}" id="theme_source">
```
and just after that line, add the following:
```
{% if site.minimal_mistakes_skin2 %}
  <link rel="stylesheet alternate" href="{{ '/assets/css/theme2.css' | relative_url }}" id="theme_source_2">
  <script>
    let theme = sessionStorage.getItem('theme');
    if(theme === "dark")
    {
      sessionStorage.setItem('theme', 'dark');
      node1 = document.getElementById('theme_source');
      node2 = document.getElementById('theme_source_2');
      node1.setAttribute('rel', 'stylesheet alternate'); 
      node2.setAttribute('rel', 'stylesheet');
    }
    else
    {
      sessionStorage.setItem('theme', 'light');
    }
  </script>
{% endif %}
```
Note that, I am using "light" and "dark" as generic names of skin1 and skin2. These strings have nothing to do with the actual skin names.

Now, create a _layouts folder and copy the _layouts/default.html from your Minimal-mistakes gems directory. Now we have to modify the file. Open it and find the following part:
```
    <div class="initial-content">
      {{ content }}
    </div>
```
Keep those lines intact and just above those, add the following:
```
    {% if site.minimal_mistakes_skin2 %}
      <button type="button" onclick="node1=document.getElementById('theme_source');node2=document.getElementById('theme_source_2');if(node1.getAttribute('rel')=='stylesheet'){node1.setAttribute('rel', 'stylesheet alternate'); node2.setAttribute('rel', 'stylesheet');sessionStorage.setItem('theme', 'dark');}else{node2.setAttribute('rel', 'stylesheet alternate'); node1.setAttribute('rel', 'stylesheet');sessionStorage.setItem('theme', 'light');} return false;">
        HI!
      </button>
    {% endif %}
```

Now if you build your site, you'll see a large button with bad design and written "HI!" on it (add your own css). Click on that to toggle theme. Also verify that the skin remains consistent if we navigate to other pages.

Here is one example, done on the theme ```basically-basic```: [https://sohamsaha99.github.io/notes/](https://sohamsaha99.github.io/notes/) . The button is at the top right.

Please suggest improvements and designs for the button.

**Reply by okielua** [reply]
@sohamsaha99 thanks for this, it works perfectly. How would I go about if I wanted the theme to stay dark when leaving and revisiting the page?

**Reply by shuds13** [reply]
I put it in the masthead (_includes/masthead.html).

**Reply by crykn** [reply]
<details>
  <summary><strike><I>If you want to automatically switch the theme according to the user's OS or browser settings, you can replace the script above by this Javascript code</I></strike></summary>
  
 ```html
<script>
     function set_theme() {
         if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
           node1 = document.getElementById('theme_source');
           node2 = document.getElementById('theme_source_2');
           node1.setAttribute('rel', 'stylesheet alternate');
           node2.setAttribute('rel', 'stylesheet');
         }
     }
     window.onload = set_theme;
</script>
```

The changes in __layouts/default.html_ are then no longer needed.

---
  
</details>

**Edit (August 2021):** @Frosty-J suggested an even simpler solution over in our website's repo (https://github.com/libgdx/libgdx.github.io/pull/45). The following CSS import query _**automatically switches the theme**_ according to the user's OS or browser settings: 

```html 
{% if site.minimal_mistakes_skin_dark %}
  <meta name="color-scheme" content="light dark">
  <style>
    @import url("{{ '/assets/css/main2.css' | relative_url }}") (prefers-color-scheme: dark);
  </style>
{% endif %}
```

The whole code (including the logo switching mentioned in the post below) can be found here: [config.yml](https://github.com/libgdx/libgdx.github.io/blob/59cc4fafd7f04755d87650f644496dde460f2c15/_config.yml#L17-L21), [_includes/masthead.html](https://github.com/libgdx/libgdx.github.io/blob/f2c59de0d644cd4865c00927ea3ca1fa1b209f74/_includes/masthead.html#L11-L21), [_includes/head.html](https://github.com/libgdx/libgdx.github.io/blob/59cc4fafd7f04755d87650f644496dde460f2c15/_includes/head.html#L21-L27), [assets/css/main2.scss](https://github.com/libgdx/libgdx.github.io/blob/59cc4fafd7f04755d87650f644496dde460f2c15/assets/css/main2.scss#L1-L8). A live example can be seen here: https://libgdx.com

**Reply by djmcgregor** [reply]
Building on this, I needed to change my masthead logo colors with the theme change. I added entry to `_config.yml` called `logo2`, following the scheme of `logo`. Then, in `masthead.html`, I added one line at the top
```liquid
{% capture logo_path2 %}{{ site.logo2 }}{% endcapture %}
```
and replaced
```liquid
{% unless logo_path == empty %}
  <a class="site-logo" href="{{ '/' | relative_url }}"><img src="{{ logo_path | relative_url }}" alt=""></a>
{% endunless %}
```
with
```liquid
{% unless logo_path == empty %}
  {% if site.minimal_mistakes_skin2 and site.logo2 %}
    <a class="site-logo" href="{{ '/' | relative_url }}">
      <picture>
        <source
          srcset="{{ logo_path2 | relative_url }}"
          media="(prefers-color-scheme: dark)"
          alt="">
        <img src="{{ logo_path | relative_url }}" alt="">
      </picture>
    </a>
  {% else %}
    <a class="site-logo" href="{{ '/' | relative_url }}"><img src="{{ logo_path | relative_url }}" alt=""></a>
  {% endif %}
{% endunless %}
```

**Reply by erikw** [reply]
@sohamsaha99 You're a star! You should make a PR to minimal-mistakes to make this an optional feature :)

**Reply by VladimirSiv** [reply]
If you want an icon next to navigation:

As @shuds13 suggested move it to `masterhead`. 

In `_includes/masterhead.html` find ` {% if site.search == true %} ` and above that add:

```
{% if site.minimal_mistakes_skin2 %}
  <i class="fas fa-fw fa-sun" aria-hidden="true" onclick="node1=document.getElementById('theme_source');node2=document.getElementById('theme_source_2');if(node1.getAttribute('rel')=='stylesheet'){node1.setAttribute('rel', 'stylesheet alternate'); node2.setAttribute('rel', 'stylesheet');sessionStorage.setItem('theme', 'dark');}else{node2.setAttribute('rel', 'stylesheet alternate'); node1.setAttribute('rel', 'stylesheet');sessionStorage.setItem('theme', 'light');} return false;"></i>
{% endif %}
```
Of course change `fa-sun` as you wish.

**Comment by fabrizioleone** [comment]
Hi,

Thanks for this, it looks very promising. The link to the example webpage is broken ( https://sohamsaha99.github.io/notes/ ). Could you please fix it to see how it looks like?

Thank you.

**Reply by VladimirSiv** [reply]
Hey @fabrizioleone 

I've added it to my website: https://www.vladsiv.com/
You can check there and also see the code here: https://github.com/VladimirSiv/VladimirSiv.github.io

**Reply by bkmgit** [reply]
This seems to work well. Perhaps updating the documentation is sufficient.

**Reply by jaketae** [reply]
I can also confirm that the above solution works well. Comparing @sohamsaha99's solution to @VladimirSiv's, I found that one should modify `masthead.html` instead of `head.html`.

**Comment by fredmayor88** [comment]
Hello, 
I'd like to achieve a following solution: 
- the website automatically detects the night mode based on system settings. (what @crykn proposed)
- but the button is still there, to give the user ability to switch the theme manually if they want (@sohamsaha99 's solution)

So basically the best of both world. 
Could someone please propose a working solution? For now I just added @crykn 's code after 
```
  <link rel="stylesheet alternate" href="{{ '/assets/css/theme2.css' | relative_url }}" id="theme_source_2">
  <script>
    let theme = sessionStorage.getItem('theme');
    if(theme === "dark")
    {
      sessionStorage.setItem('theme', 'dark');
      node1 = document.getElementById('theme_source');
      node2 = document.getElementById('theme_source_2');
      node1.setAttribute('rel', 'stylesheet alternate'); 
      node2.setAttribute('rel', 'stylesheet');
    }
    else
    {
      sessionStorage.setItem('theme', 'light');
    }
  </script>
```

and there is one problem: when the OS theme is set to dark, the theme change button on the website doesn't cause any effects.
Thanks a lot for a great thread everyone.

**Comment by mmistakes** [comment]
Untested code but I think if you modify the `theme` conditional to check if it's set to `dark` or if a user's system color scheme prefers dark you should get the results you want.

```diff
 <link rel="stylesheet alternate" href="{{ '/assets/css/theme2.css' | relative_url }}" id="theme_source_2">
  <script>
    let theme = sessionStorage.getItem('theme');
-    if(theme === 'dark')
+    if (theme === 'dark' || window.matchMedia('(prefers-color-scheme: dark)').matches)
    {
      sessionStorage.setItem('theme', 'dark');
      node1 = document.getElementById('theme_source');
      node2 = document.getElementById('theme_source_2');
      node1.setAttribute('rel', 'stylesheet alternate'); 
      node2.setAttribute('rel', 'stylesheet');
    }
    else
    {
      sessionStorage.setItem('theme', 'light');
    }
  </script>
```

**Reply by fredmayor88** [reply]
@mmistakes thanks a lot, that was fast! But it doesn't work :(

**Comment by uzza1hossain** [comment]
I implemented night mode using this method. It works fine. Thanks.

But I am facing an issue with giscus comment provider. How can I toggle giscus theme with site theme?

**Comment by codeslake** [comment]
I love the solution provided by @sohamsaha99, Thanks!!

But sometimes, whenever I reload a page or change the theme, CSS is loaded after HTML (I can see my form fields for a brief moment while they are still unstyled).

I think I've solved this by first loading `link`s for stylesheets with `rel="stylesheet"` (this solved the CSS issues when reloading the page for me):
```html
// in head.html
<link rel="stylesheet" href="{{ '/assets/css/main.css' | relative_url }}" id="theme_source">
{% if site.minimal_mistakes_skin2 %}
  <link rel="stylesheet" href="{{ '/assets/css/theme2.css' | relative_url }}" id="theme_source_2">
  <script>
    if (sessionStorage.getItem('theme') === null) {
      if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        sessionStorage.setItem('theme', 'dark');
      }
      else{
        sessionStorage.setItem('theme', 'light');
      }
    }

    let theme = sessionStorage.getItem('theme');
    if(theme === "dark"){
      document.getElementById('theme_source').setAttribute('rel', 'stylesheet alternate'); 
      sessionStorage.setItem('theme', 'dark');
    }
    else
    {
      document.getElementById('theme_source_2').setAttribute('rel', 'stylesheet alternate'); 
      sessionStorage.setItem('theme', 'light');
    }
  </script>
{% endif %}
```

And for the button script, I've set a timeout for disabling a stylesheet (this solved the CSS issues when applying theme change for me):
```html
// in _includes/masthead.html
<div class="masthead">
     .....
          {% if site.minimal_mistakes_skin2 %}
            <li><i class="fas fa-fw fa-adjust dark_mode_btn" aria-hidden="true" onclick="dark_mode_btn_click()"></i></li>
          {% endif %}
     .....
</div>

// in _includes/scripts.html
<script>
  function dark_mode_btn_click() {
    var node1 = document.getElementById('theme_source');
    var node2 = document.getElementById('theme_source_2');
    if(node1.getAttribute('rel')=='stylesheet'){
      node2.setAttribute('rel', 'stylesheet');
      setTimeout(function(){
        node1.setAttribute('rel', 'stylesheet alternate');
      }, 10);
      sessionStorage.setItem('theme', 'dark');
    }else{
      node1.setAttribute('rel', 'stylesheet');
      setTimeout(function(){
        node2.setAttribute('rel', 'stylesheet alternate');
      }, 10);
      sessionStorage.setItem('theme', 'light');
    }
    return false;
}
<script>
```

**Reply by segunak** [reply]
> I love the solution provided by @sohamsaha99, Thanks!!
> 
> But sometimes, whenever I reload a page or change the theme, CSS is loaded after HTML (I can see my form fields for a brief moment while they are still unstyled).
> 
> I think I've solved this by first loading `link`s for stylesheets with `rel="stylesheet"` (this solved the CSS issues when reloading the page for me):
> 
> ```
> // in head.html
> <link rel="stylesheet" href="{{ '/assets/css/main.css' | relative_url }}" id="theme_source">
> {% if site.minimal_mistakes_skin2 %}
>   <link rel="stylesheet" href="{{ '/assets/css/theme2.css' | relative_url }}" id="theme_source_2">
>   <script>
>     if (sessionStorage.getItem('theme') === null) {
>       if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
>         sessionStorage.setItem('theme', 'dark');
>       }
>       else{
>         sessionStorage.setItem('theme', 'light');
>       }
>     }
> 
>     let theme = sessionStorage.getItem('theme');
>     if(theme === "dark"){
>       document.getElementById('theme_source').setAttribute('rel', 'stylesheet alternate'); 
>       sessionStorage.setItem('theme', 'dark');
>     }
>     else
>     {
>       document.getElementById('theme_source_2').setAttribute('rel', 'stylesheet alternate'); 
>       sessionStorage.setItem('theme', 'light');
>     }
>   </script>
> {% endif %}
> ```
> 
> And for the button script, I've set a timeout for disabling a stylesheet (this solved the CSS issues when applying theme change for me):
> 
> ```
> // in _includes/masthead.html
> <div class="masthead">
>      .....
>           {% if site.minimal_mistakes_skin2 %}
>             <li><i class="fas fa-fw fa-adjust dark_mode_btn" aria-hidden="true" onclick="dark_mode_btn_click()"></i></li>
>           {% endif %}
>      .....
> </div>
> 
> // in _includes/scripts.html
> <script>
>   function dark_mode_btn_click() {
>     var node1 = document.getElementById('theme_source');
>     var node2 = document.getElementById('theme_source_2');
>     if(node1.getAttribute('rel')=='stylesheet'){
>       node2.setAttribute('rel', 'stylesheet');
>       setTimeout(function(){
>         node1.setAttribute('rel', 'stylesheet alternate');
>       }, 10);
>       sessionStorage.setItem('theme', 'dark');
>     }else{
>       node1.setAttribute('rel', 'stylesheet');
>       setTimeout(function(){
>         node2.setAttribute('rel', 'stylesheet alternate');
>       }, 10);
>       sessionStorage.setItem('theme', 'light');
>     }
>     return false;
> }
> <script>
> ```

Was seeing the same issue where reloading the page would have little moments where things didn't look right. This solved it for me. Great work!

**Reply by rnag** [reply]
Thanks, above solution from @codeslake worked for me.

I had another question, and it may seem dumb. I have some HTML elements where I need to bring in different CSS classes (different image from FA basically) when it is dark mode instead of light. Anyone have an idea how to achieve this?

**Comment by amomchilov** [comment]
Piling on the list of "here's what I tried that worked", I started using the trick written up by @codeslake, @codeslake and [documented](https://humberto.stein-shiromoto.net/posts/2022/09/27/blog-post_adding_theme_switch_to_minimal_mistakes) by @hsteinshiromoto. Instead of having an explicit button, I just wanted to use whichever preference the user has set on their OS.

Turns out that you can [add media queries](https://stackoverflow.com/a/70516556/3141234) to `<link>` tags, so it was as simple as adding this second stylesheet in `_includes/head.html`:

```html
<link title="main" rel="stylesheet" href="{{ '/assets/css/main.css' | relative_url }}" >
<link title="main" rel="stylesheet" href="{{ '/assets/css/dark.css' | relative_url }}" media="(prefers-color-scheme: dark)">
```

**Reply by akashmjn** [reply]
Came across this, and worked great - thanks!

Using a media query to link compiled `css` files gets rid of most of the complexity (buttons etc) from @sohamsaha99 's proposal in https://github.com/mmistakes/minimal-mistakes/discussions/2033#discussioncomment-257421. 

In summary:
- `main.scss` ultimately gets pre-processed by jekyll into `_site/assets/css/main.css` 
- this css is linked to output static html pages controlled by `_includes/head.html`
- the solution above gets you to create a second `dark.css` that so it can be linked conditionally based on the OS

Simple and likely covers 90% of what most need 👌

**Comment by crykn** [comment]
@amomchilov Adding the media query to the `<link>` tag is pretty nifty! I might add that to the example in [my comment above](https://github.com/mmistakes/minimal-mistakes/discussions/2033#discussioncomment-460914) which also includes optional automatic logo switching.

**Reply by amomchilov** [reply]
@crykn I haven't bothered to implement a clickable toggle (I figure matching the OS pref is prob want 99% of people want anyway), but it might be interesting to combine these two techniques.

So the media query is used to set the default (and respond to changes in the OS pref, e.g. at sunset), while the button can be used to override it.

**Reply by crykn** [reply]
The solution in my comment doesn't use a button either; it also depends solely on the OS settings.

**Comment by sorat0mo** [comment]
Came across this thread, also looked for ways to toggle giscus themes along with dark mode. I'll help everyone else summarize what needs to be done. I referred to this when implementing my giscus toggle: https://github.com/giscus/giscus/issues/336#issuecomment-1007922777

<details>
<summary> The full code in action:</summary>

![](https://files.catbox.moe/zhm3xc.gif)
</details>

I am a noob in javascript, so I combined some of the above into a new js asset:
<details>

<summary>dark-mode.js</summary>

```javascript
function changeGiscusTheme(theme) {
      function sendMessage(message) {
            const iframe = document.querySelector('iframe.giscus-frame');
            if (!iframe) return;
            iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
            setConfig: {
                  theme: theme
            }
      });
}

function updateThemeButton(){
      let currentTheme = sessionStorage.getItem('theme');
      let themeToggle = document.getElementById('theme-toggle');
      if (currentTheme === 'dark') {
            themeToggle.classList.remove('fa-adjust');
            themeToggle.classList.add('fa-sun');
      } else {
            themeToggle.classList.remove('fa-sun');
            themeToggle.classList.add('fa-adjust');
      }
}

function themeOnLoad(){
let theme = sessionStorage.getItem('theme');
    if(theme === "dark" || (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches && !(sessionStorage.getItem('manualtoggled'))))
    {
      sessionStorage.setItem('theme', 'dark');
      node1 = document.getElementById('theme_source');
      node2 = document.getElementById('theme_source_2');
      node1.setAttribute('rel', 'stylesheet alternate'); 
      node2.setAttribute('rel', 'stylesheet');
    }
    else
    {
      sessionStorage.setItem('theme', 'light');
    }
    updateThemeButton();
}

function toggleTheme() {
      node1 = document.getElementById('theme_source');
      node2 = document.getElementById('theme_source_2');
      if (node1.getAttribute('rel')=='stylesheet') {
            node1.setAttribute('rel', 'stylesheet alternate');
            node2.setAttribute('rel', 'stylesheet');
            sessionStorage.setItem('theme', 'dark');
            changeGiscusTheme("dark");
          } else {
            node1.setAttribute('rel', 'stylesheet');
            node2.setAttribute('rel', 'stylesheet alternate');
            sessionStorage.setItem('theme', 'light');
            changeGiscusTheme("light");
          }
      updateThemeButton();
      sessionStorage.setItem('manualtoggled', true);
}
```
</details>

Save the above code somewhere in the repo directory. *Remove the `changeGiscusTheme` sections if you do not use Giscus*.

Add the follwing in `_config.yaml`. Uncomment both logos if you do use logo. I don't have one so I commented them out: 
```yaml
minimal_mistakes_skin_dark: "dark"

# logo: "/assets/brand/logo.svg"
# logo_dark: "/assets/brand/logo_dark.svg"
head_scripts:
  - /assets/js/dark-mode.js
 ```

Then, in `masthead.html`, I added the following above `{% if site.search == true %}`:
```liquid
<i class="fas fa-fw fa-adjust" id="theme-toggle" aria-hidden="true" title="Toggle between light and dark mode" onclick="toggleTheme(); return false;"></i>
<script>themeOnLoad();</script>
```

Also, in `masthead.html`, find the following lines:
```liquid
{% unless logo_path == empty %}
  <a class="site-logo" href="{{ '/' | relative_url }}"><img src="{{ logo_path | relative_url }}" alt="{{ site.masthead_title | default: site.title }}"></a>
{% endunless %}
```

Replace the above with the following:
```liquid
{% unless logo_path == empty %}
  <a class="site-logo" href="{{ '/' | relative_url }}">
      <picture>
        {% unless logo_path_dark == empty %}
          <!-- dark mode logo -->
          <source
            srcset="{{ logo_path_dark | relative_url }}"
            media="(prefers-color-scheme: dark)">
        {% endunless %}
        <img src="{{ logo_path | relative_url }}" alt="{{ site.masthead_title | default: site.title }}">
      </picture>
  </a>
{% endunless %}
```

In `head.html`, replace `<link rel="stylesheet" href="/assets/css/main.css">` with the following:
```html
<link title="main" rel="stylesheet" href="{{ '/assets/css/main.css' | relative_url }}" id="theme_source">
<link rel="stylesheet alternate" href="{{ '/assets/css/main2.css' | relative_url }}" id="theme_source_2">
```

Create a new file `/assets/css/main2.css` with the folllowing content:
```css
@charset "utf-8";

@import "minimal-mistakes/skins/{{ site.minimal_mistakes_skin_dark | default: 'dark' }}"; // skin
@import "minimal-mistakes"; // main partials
```

### If you use Giscus
When loading into a new post or page with giscus enabled, the theme will be set by default. Since we set the value `theme` via sessionstorage, we can directly use that to set our giscus theme, so that the giscus theme will follow our preference on loading a new page. 

Find the following `var script` line in `_includes/comments-providers/giscus.html` and add the last line (copy and modify if you are using remote-theme):
```javascript
var script = document.createElement('script');
let giscusTheme = sessionStorage.getItem("theme");
```

Then, find the `setAttribute` line below and change it accordingly:
```javascript
script.setAttribute('data-theme', giscusTheme);
```

Note: I am using the manual toggle solution provided by https://github.com/mmistakes/minimal-mistakes/discussions/2033#discussioncomment-257421, you also need to check the names of your variables. Not sure if the above function has redundant parts, but I am happy with the outcome for the time being.

You can see the effect in action here: https://lilme.cc/new-year-old-challenge/

Edit: After a lot of head-scratching and edits, I believe this is the final version that works and looks good. Hopefully dark mode toggle will be implemented in the official theme in the future.
Edit 2: I decided to tweak this a bit. Now this script respects 'prefers-color-scheme' on load, and will only default to light theme if the user explicitly switched to a light theme. Works nicely on my website, a fun little experiment for a javascript noob.

**Comment by kamalx** [comment]
Hi, since this discussion appears to be open and I landed up here while looking for reference for the same native support for basic theme toggle that the OP/OC was asking about, which is available now:

https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/color_value/light-dark

```css
:root {
  color-scheme: light dark;
}

/* you can define the variables in the above block to use them */
/* tweak the below selector to be any specificity you require */
html * {
  color: light-dark(var(--light, #222), var(--dark, #89A));
  background-color: light-dark(var(--light-bg, #ccc), var(--dark-bg,#222));
}

/*
in its current form, without the variables defined, the above lines are equivalent to writing just this:

html * {
  color: light-dark(#222, #89A);
  background-color: light-dark(#ccc, #222);
}
*/
```

Here's a screenshot of the docs with the above rules applied, which will change back to the original light theme if i change my system theme to light:

<img width="1091" height="440" alt="image" src="https://github.com/user-attachments/assets/68315b15-5741-4d1a-a3e5-e5942fac714f" />
